### ISSUE TITLE
git.detached does not fetch if rev is missing from local

### ISSUE THREAD

**Issue by abulford** [saltstack/salt#43143]
### Description of Issue/Question
When applying the `git.detached` state with a `rev` specified, if the `rev` is set to a commit hash which is available on the remote but not locally, the state fails with an exception.

### Setup
Example state file:
```
# git-state.sls
https://github.com/saltstack/salt.git:
  git.detached:
    - rev: '5cb40aca4269174ff633f66ec76fb56b6880d2ae'
    - target: '/var/lib/salt/git-clones/salt'
```

Repo has already been cloned to the target but a new commit has been added to the remote since the clone, the state is specifying that new commit hash in `rev`.

### Steps to Reproduce Issue
With the setup above, apply state:
```
# salt-call -l debug state.apply master.external-modules
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/_schedule.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/beacons.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/beacons.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/engine.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/engine.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/f_defaults.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/f_defaults.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/reactor.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/reactor.conf
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: <host>.<domain>
[DEBUG   ] Configuration file path: /etc/salt/minion
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/_schedule.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/beacons.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/beacons.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/engine.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/engine.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/f_defaults.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/f_defaults.conf
[DEBUG   ] Including configuration from '/etc/salt/minion.d/reactor.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/reactor.conf
[DEBUG   ] Please install 'virt-what' to improve results of the 'virtual' grain.
[DEBUG   ] Connecting to master. Attempt 1 of 1
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', '<host>.<domain>', 'tcp://127.0.0.1:4506')
[DEBUG   ] Generated random reconnect delay between '1000ms' and '11000ms' (1030)
[DEBUG   ] Setting zmq_reconnect_ivl to '1030ms'
[DEBUG   ] Setting zmq_reconnect_ivl_max to '11000ms'
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', '<host>.<domain>', 'tcp://127.0.0.1:4506', 'clear')
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] SaltEvent PUB socket URI: /var/run/salt/minion/minion_event_d9754c0b30_pub.ipc
[DEBUG   ] SaltEvent PULL socket URI: /var/run/salt/minion/minion_event_d9754c0b30_pull.ipc
[DEBUG   ] Initializing new IPCClient for path: /var/run/salt/minion/minion_event_d9754c0b30_pull.ipc
[DEBUG   ] Sending event: tag = salt/auth/creds; data = {'_stamp': '2017-08-23T13:32:48.753947', 'creds': {'publish_port': 4505, 'aes': '<redacted>', 'master_uri': 'tcp://127.0.0.1:4506'}, 'key': ('/etc/salt/pki/minion', '<host>.<domain>', 'tcp://127.0.0.1:4506')}
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] Determining pillar cache
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', '<host>.<domain>', 'tcp://127.0.0.1:4506', 'aes')
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', '<host>.<domain>', 'tcp://127.0.0.1:4506')
[DEBUG   ] Loaded minion key: /etc/salt/pki/minion/minion.pem
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] LazyLoaded state.apply
[DEBUG   ] LazyLoaded saltutil.is_running
[DEBUG   ] LazyLoaded grains.get
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', '<host>.<domain>', 'tcp://127.0.0.1:4506', 'aes')
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', '<host>.<domain>', 'tcp://127.0.0.1:4506')
[INFO    ] Loading fresh modules for state activity
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] In saltenv 'base', looking at rel_path 'git-state.sls' to resolve 'salt://git-state.sls'
[DEBUG   ] In saltenv 'base', ** considering ** path '/var/cache/salt/minion/files/base/git-state.sls' to resolve 'salt://git-state.sls'
[DEBUG   ] compile template: /var/cache/salt/minion/files/base/git-state.sls
[DEBUG   ] Jinja search path: ['/var/cache/salt/minion/files/base']
[PROFILE ] Time (in seconds) to render '/var/cache/salt/minion/files/base/git-state.sls' using 'jinja' renderer: 0.00683093070984
[DEBUG   ] Rendered data from file: /var/cache/salt/minion/files/base/git-state.sls:
# git-state.sls
https://github.com/saltstack/salt.git:
  git.detached:
    - rev: '5cb40aca4269174ff633f66ec76fb56b6880d2ae'
    - target: '/var/lib/salt/git-clones/salt'
[DEBUG   ] LazyLoaded config.get
[DEBUG   ] Results of YAML rendering: 
OrderedDict([('https://github.com/saltstack/salt.git', OrderedDict([('git.detached', [OrderedDict([('rev', '5cb40aca4269174ff633f66ec76fb56b6880d2ae')]), OrderedDict([('target', '/var/lib/salt/git-clones/salt')])])]))])
[PROFILE ] Time (in seconds) to render '/var/cache/salt/minion/files/base/git-state.sls' using 'yaml' renderer: 0.0134608745575
[DEBUG   ] LazyLoaded git.version
[DEBUG   ] LazyLoaded cmd.run_all
[INFO    ] Executing command ['git', '--version'] in directory '/root'
[DEBUG   ] stdout: git version 1.8.3.1
[DEBUG   ] LazyLoaded git.detached
[INFO    ] Running state [https://github.com/saltstack/salt.git] at time 13:32:49.637751
[INFO    ] Executing state git.detached for [https://github.com/saltstack/salt.git]
[INFO    ] Checking local revision for /var/lib/salt/git-clones/salt
[INFO    ] Executing command ['git', 'rev-parse', 'HEAD'] in directory '/var/lib/salt/git-clones/salt'
[DEBUG   ] stdout: 1280c03c31599ed89d755277478bdcf79b03befb
[INFO    ] Checking local branch for /var/lib/salt/git-clones/salt
[INFO    ] Executing command ['git', 'rev-parse', '--abbrev-ref', 'HEAD'] in directory '/var/lib/salt/git-clones/salt'
[DEBUG   ] stdout: HEAD
[INFO    ] Executing command ['git', 'describe', '--always', '5cb40aca4269174ff633f66ec76fb56b6880d2ae'] in directory '/var/lib/salt/git-clones/salt'
[ERROR   ] Command '['git', 'describe', '--always', '5cb40aca4269174ff633f66ec76fb56b6880d2ae']' failed with return code: 128
[ERROR   ] stderr: fatal: 5cb40aca4269174ff633f66ec76fb56b6880d2ae is not a valid 'commit' object
[ERROR   ] retcode: 128
[ERROR   ] An exception occurred in this state: Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/state.py", line 1837, in call
    **cdata['kwargs'])
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1794, in wrapper
    return f(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/states/git.py", line 2236, in detached
    password=password):
  File "/usr/lib/python2.7/site-packages/salt/modules/git.py", line 1557, in describe
    ignore_retcode=ignore_retcode)['stdout']
  File "/usr/lib/python2.7/site-packages/salt/modules/git.py", line 320, in _git_run
    raise CommandExecutionError(msg)
CommandExecutionError: Command 'git describe --always 5cb40aca4269174ff633f66ec76fb56b6880d2ae' failed: fatal: 5cb40aca4269174ff633f66ec76fb56b6880d2ae is not a valid 'commit' object

[INFO    ] Completed state [https://github.com/saltstack/salt.git] at time 13:32:49.744547 duration_in_ms=106.805
[DEBUG   ] File /var/cache/salt/minion/accumulator/62923920 does not exist, no need to cleanup.
[DEBUG   ] LazyLoaded config.option
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', '<host>.<domain>', 'tcp://127.0.0.1:4506', 'aes')
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', '<host>.<domain>', 'tcp://127.0.0.1:4506')
[DEBUG   ] LazyLoaded highstate.output
local:
----------
          ID: https://github.com/saltstack/salt.git
    Function: git.detached
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib/python2.7/site-packages/salt/state.py", line 1837, in call
                  **cdata['kwargs'])
                File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1794, in wrapper
                  return f(*args, **kwargs)
                File "/usr/lib/python2.7/site-packages/salt/states/git.py", line 2236, in detached
                  password=password):
                File "/usr/lib/python2.7/site-packages/salt/modules/git.py", line 1557, in describe
                  ignore_retcode=ignore_retcode)['stdout']
                File "/usr/lib/python2.7/site-packages/salt/modules/git.py", line 320, in _git_run
                  raise CommandExecutionError(msg)
              CommandExecutionError: Command 'git describe --always 5cb40aca4269174ff633f66ec76fb56b6880d2ae' failed: fatal: 5cb40aca4269174ff633f66ec76fb56b6880d2ae is not a valid 'commit' object
     Started: 13:32:49.637742
    Duration: 106.805 ms
     Changes:   

Summary for local
------------
Succeeded: 0
Failed:    1
------------
Total states run:     1
Total run time: 106.805 ms
```

### Versions Report
```
$ salt --versions-report
Salt Version:
           Salt: 2017.7.1
 
Dependency Versions:
           cffi: 1.6.0
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: 1.10.6
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.9.6
        libgit2: 0.24.6
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
   pycryptodome: 3.4.3
         pygit2: 0.24.2
         Python: 2.7.5 (default, Nov  6 2016, 00:28:07)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.6
 
System Versions:
           dist: centos 7.3.1611 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-514.16.1.el7.x86_64
         system: Linux
        version: CentOS Linux 7.3.1611 Core
```

**Comment by abulford** [comment]
Having looked into the state module, it seems the logic is there to perform a fetch if the commit isn't found locally, the problem is that the `git.describe` call throws an exception rather than returning false.

This could be fixed by exception handling in the state module's `detached` function, exception handling in the execution module's `describe` function, or by passing `failhard=False` to `_git_run` in the `describe` function.  I don't mind submitting a PR with a fix if it can be agreed what the best approach is.

**Comment by Ch3LL** [comment]
@terminalmage can you help give some insight to @abulford as to what the best approach to fix this is?

**Comment by terminalmage** [comment]
I decided to fix it myself, because I noticed that there was a variable naming change between 2016.11 and 2017.7, and I wanted to make sure there was no possibility of a bug creeping in on a merge-forward.

@abulford See https://github.com/saltstack/salt/pull/43178/files. This is similar to what is already being done in the `git.latest` state [here](https://github.com/saltstack/salt/blob/2016.11/salt/states/git.py#L939-L948) to check if the local checkout has the desired rev.

**Comment by abulford** [comment]
@terminalmage I gave it a run and found a couple of issues which I've pointed out in the PR, but otherwise it seems to do the trick :).

**Comment by terminalmage** [comment]
@abulford thanks, I addressed both of the issues you found.

**Comment by abulford** [comment]
@terminalmage no worries :+1:

**Comment by terminalmage** [comment]
This was fixed in #43178.
